### PR TITLE
feat(developer): ldml scan codes support 🙀 

### PR DESCRIPTION
**Pull request by srl295**
- pivot:  this now dynamically generates the vkey map based on CLDR scan codes
- add infrastructure to detect imported items
- warn (hint) if user adds a custom scancode
- also includes an LDML resource update  - including removal of the `vkeys`, see #7135 
- also minor, use source maps when launching `kmc` from tests.

Fixes: #9403 
@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/types/src/consts/virtual-key-constants.ts [hunk 1]
```diff
@@ -174,13 +187,13 @@ export const ISOVirtualKeyMap: KeyMap = [
 
 export const JISVirtualKeyMap: KeyMap = [
   // [Hankaku/Zenkaku] 1 2 3 4 5 6 7 8 9 0 - ^ ¥ [bksp]
-  [ k.K_1, k.K_2, k.K_3, k.K_4, k.K_5, k.K_6, k.K_7, k.K_8, k.K_9, k.K_0, k.K_HYPHEN, k.K_EQUAL, k.K_BKSLASH /* YEN */ ],
+  [ k.K_BKQUOTE, k.K_1, k.K_2, k.K_3, k.K_4, k.K_5, k.K_6, k.K_7, k.K_8, k.K_9, k.K_0, k.K_HYPHEN, k.K_EQUAL, k.K_BKSLASH /* YEN */ ],
   // [tab] Q W E R T Y U I O P @«`» [ [enter]
```
Comment (mcdurdin) [inline common/web/types/src/consts/virtual-key-constants.ts:191]:
```suggestion
  // [tab] Q W E R T Y U I O P @ [ [enter]
```

File: common/web/types/src/consts/virtual-key-constants.ts [hunk 2]
```diff
@@ -174,13 +187,13 @@ export const ISOVirtualKeyMap: KeyMap = [
...
   [ k.K_Q, k.K_W, k.K_E, k.K_R, k.K_T, k.K_Y, k.K_U, k.K_I, k.K_O, k.K_P, k.K_BKQUOTE, k.K_LBRKT  ],
   // [caps] A S D F G H J K L ; : ] [enter]
   [ k.K_A, k.K_S, k.K_D, k.K_F, k.K_G, k.K_H, k.K_J, k.K_K, k.K_L, k.K_COLON, k.K_QUOTE, k.K_RBRKT ],
   // [shift] Z X C V B N M , . / _ [shift]
-  [ k.K_Z, k.K_X, k.K_C, k.K_V, k.K_B, k.K_N, k.K_M, k.K_COMMA, k.K_PERIOD, k.K_SLASH, k.K_oE2 /* ろ */ ],
+  [ k.K_oE2, k.K_Z, k.K_X, k.K_C, k.K_V, k.K_B, k.K_N, k.K_M, k.K_COMMA, k.K_PERIOD, k.K_SLASH, k.k_oC1 /* ろ */ ],
```
Comment (mcdurdin) [inline common/web/types/src/consts/virtual-key-constants.ts:196]: My JIS keyboard doesn't have a key between Shift and Z:



FWIW, in Windows, the virtual key code at far-right (scan code 0x73) is `K_oE2` when using Japanese base keyboard, or `K_oC1` when using US base keyboard. Yay VK!

We should probably use `K_oC1` internally for scan code 0x73 in Keyman Engine for Windows, which we need to track as a separate issue. See also the KS discussion above.

File: common/web/types/src/ldml-keyboard/ldml-keyboard-xml.ts [hunk 3]
```diff
@@ -193,3 +194,36 @@ export interface LKDisplays {
...
+};
+
+/**
+ * Utilities for determining the import status of items
+ */
+export class ImportStatus {
+  /** item came in via implied (spec based) import, such as keys-Latn-implied.xml */
+  static impliedImport = Symbol.for('@keymanapp:implied_import');
+  /** item came in via import */
+  static import = Symbol.for('@keymanapp:import');
```
Comment (mcdurdin) [inline common/web/types/src/ldml-keyboard/ldml-keyboard-xml.ts:218]: This introduces a new namespace that we need to be careful with -- the `@keymanapp` Symbol namespace. Conceptually I think it is fine but I am just a little wary about future trouble with it. Might be good to use a longer prefix, for disambiguation e.g.:

```suggestion
  static impliedImport = Symbol.for('@keymanapp:common/ldml/implied_import');
  /** item came in via import */
  static import = Symbol.for('@keymanapp:common/ldml/import');
```

Also, just verifying is this just a way of tagging arbitrary structures with an "imported" flag? And you wanted to use this to avoid potential stumbling blocks in the way we use those structures, because it won't be an interable key on an object?

File: developer/src/kmc-ldml/src/compiler/messages.ts [hunk 4]
```diff
@@ -149,5 +149,9 @@ export class CompilerMessages {
   m(this.ERROR_DisplayNeedsToOrId, `display ${CompilerMessages.toOrId(o)} needs to= or id=, but not both`);
   static ERROR_DisplayNeedsToOrId = SevError | 0x0022;
 
+  static Hint_UnsupportedCustomForm = (o:{id: string}) =>
+  m(this.HINT_UnsupportedCustomForm, `Custom <form id="${o.id}"> element is not supported.`);
+  static HINT_UnsupportedCustomForm = SevHint | 0x0023;
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/messages.ts:154]: Should this be a hint or a warning? Does the custom form cause breaking issues down the line?

File: developer/src/kmc-ldml/test/fixtures/sections/layr/hint-custom-form.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    <key id="grave" to="`" />
+    <key id="one" to="1" />
+  </keys>
+  <forms>
+    <!-- warning, custom scancodes -->
+    <form id="us">
+      <scanCodes codes="01 02 03 04" />
+    </form>
+  </forms>
+  <layers form="us"> <!-- INVALID: 'holographic' is not a valid form. -->
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/test/fixtures/sections/layr/hint-custom-form.xml:19]: I don't understand the comment here

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Not quite ready to approve this, sorry!

File: common/web/types/src/ldml-keyboard/ldml-keyboard-xml.ts (hunk 3 above)
Comment (srl295) [reply common/web/types/src/ldml-keyboard/ldml-keyboard-xml.ts:218]: On reflection, I think this doesn't need a named symbol at all. this symbol should be used only within one run of kmc.  it doesn't need to be independently detected. I will drop the namespace.

yes, it's just a way to tag the structure as imported - it's not iterable on the object, it doesn't affect the API at all at any level. It's metadata answering the question "how did this data object come in?"

File: developer/src/kmc-ldml/test/fixtures/sections/layr/hint-custom-form.xml (hunk 5 above)
Comment (srl295) [reply developer/src/kmc-ldml/test/fixtures/sections/layr/hint-custom-form.xml:19]: copy and paste, fixed

File: developer/src/kmc-ldml/src/compiler/messages.ts (hunk 4 above)
Comment (srl295) [reply developer/src/kmc-ldml/src/compiler/messages.ts:154]: maybe a warning is better than a hint. Basically, it's not processed by kmc.

File: common/web/types/src/consts/virtual-key-constants.ts (hunk 2 above)
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts:196]: OK, mine didn't have one there either. Looks like a copy and paste. thanks

Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts:196]: This is broken in the CLDR file, so opened a linked PR there.

File: resources/standards-data/ldml-keyboards/techpreview/import/scanCodes-implied.xml
Comment (srl295) [inline resources/standards-data/ldml-keyboards/techpreview/import/scanCodes-implied.xml:1]: (This is a file comment. Somehow, I haven't seen that feature before.)

i'm updating this prospectively from the parallel CLDR PR.

**Comment by srl295** [comment]
@mcdurdin ptal

File: common/web/types/src/consts/virtual-key-constants.ts [hunk 6]
```diff
@@ -159,6 +159,19 @@ export const USVirtualKeyMap: KeyMap = [
   [ k.K_SPACE ],
 ];
 
+export const KSVirtualKeyMap: KeyMap = [
+  // ` 1 2 3 4 5 6 7 8 9 0 - = [bksp]
+  [ k.K_BKQUOTE, k.K_1, k.K_2, k.K_3, k.K_4, k.K_5, k.K_6, k.K_7, k.K_8, k.K_9, k.K_0, k.K_HYPHEN, k.K_EQUAL, k.K_BKSLASH /* YEN */  ],
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/types/src/consts/virtual-key-constants.ts:164]: > On a Japanese keyboard, 0x7D produces `K_BKSLASH`, and 0x2B produces `K_oC1`. That may be the safest pattern for us to match.

And then I realised that that won't work. 0x2B is `K_BKSLASH` on kbdus. So we'll need to use something else for 0x7D (`K_oC1`?) in order to differentiate, because we can't change the already assigned 0x2B. The actual code doesn't really matter all that much so long as we are consistent internally.

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/types/src/consts/virtual-key-constants.ts:164]: So I am wondering about the advisability of these layout maps, as they form a separate source of overlapping truth -- we already have scanCodes-implied.xml. I think we should just a have single scancode:vk map somewhere, and construct layout maps from that+scanCodes-implied.xml.

[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts:164]: could it be `k.K_oE2` ? which is sometimes backslash-like things in ISO and ABNT2

[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts:164]: @mcdurdin  scancodes to vkeys… seems to make sense, should I just build up a table from the data we already have here? (maybe merge this first as a baseline)

On the maybe-plus side, that means we actually could support custom scancodes! Did I say that out loud?

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/types/src/consts/virtual-key-constants.ts]: `K_oE2` is mapped to scan code 0x56. I don't think we should use the same VK for different scan codes.

Internally we probably could have been using scan codes directly rather than VKs. But that boat has sailed.

[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts]: okay.  I'll keep looking.

it sounds like there should also be a test,  a sort of virtual-`@mcdurdin`, which verifies 1:1 mapping from scan code to vkey here.

File: common/web/types/src/consts/virtual-key-constants.ts [hunk 7]
```diff
@@ -159,6 +159,19 @@ export const USVirtualKeyMap: KeyMap = [
   [ k.K_SPACE ],
 ];
 
+export const KSVirtualKeyMap: KeyMap = [
+  // ` 1 2 3 4 5 6 7 8 9 0 - = YEN [bksp]
+  [ k.K_BKQUOTE, k.K_1, k.K_2, k.K_3, k.K_4, k.K_5, k.K_6, k.K_7, k.K_8, k.K_9, k.K_0, k.K_HYPHEN, k.K_EQUAL, k.K_oE2 /* YEN */  ],
+  // [tab] Q W E R T Y U I O P [ ] \
+  [ k.K_Q, k.K_W, k.K_E, k.K_R, k.K_T, k.K_Y, k.K_U, k.K_I, k.K_O, k.K_P, k.K_LBRKT, k.K_RBRKT, k.k_oC1 ],
```
[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts:166]: Given the upstream PR. 

Think I may attempt the dynamic approach. Or at least generated (for perspicuity).

[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/web/types/src/consts/virtual-key-constants.ts:166]: Dynamic sounds like a great idea. More front-end work but longer term much cleaner.

**Comment by srl295** [comment]
i'm temporarily skipping the scancode test— to show that the LDML updates are otherwise passing.

**Comment by srl295** [comment]
temporary file output => https://gist.github.com/srl295/48685c96bb23ee19ce1f14428ed5d91c

```json
 {
  "scan": "0A",
  "vname": "K_9",
  "vcode": 57
 },
 {
  "scan": "0B",
  "vname": "K_0",
  "vcode": 48
 },
 {
  "scan": "0C",
  "vname": "K_HYPHEN",
  "vcode": 189
 },
```

**Comment by mcdurdin** [comment]
That works, a bit verbose, but certainly easy to work with. It will be a separate generated file?

Evil alternative:

```javascript
const scancodes = [
  ["0A","K_9",57],
  ["0B","K_0",48],
  ["0C","K_HYPHEN",189]
].map(x => ({scan:x[0], vname:x[1], vcode:x[2]}));
```

File: common/web/types/src/consts/virtual-key-constants.ts [hunk 8]
```diff
@@ -174,13 +187,13 @@ export const ISOVirtualKeyMap: KeyMap = [
 
 export const JISVirtualKeyMap: KeyMap = [
```
[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts]: ptal at the new approach…

File: common/web/types/src/ldml-keyboard/ldml-keyboard-xml.ts (hunk 3 above)
Comment (srl295) [reply common/web/types/src/ldml-keyboard/ldml-keyboard-xml.ts]: OK here?

File: common/web/types/src/consts/virtual-key-constants.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts]: that's now given by definition…

**Comment by srl295** [comment]
@mcdurdin PTAL at progress but TODO:
- [ ] unit test: verify that scan code map is 1:1
- [ ] kmc: error if CLDR scancode results in no vkey

Is there any other checking needed?

**Comment by srl295** [comment]
```
  0x56: k.K_oE2, // << Dup
  0x73: k.k_oC1,
  0x7D: k.K_oE2, // << Dup
```

except we already have dups.

**Comment by srl295** [comment]
Okay, PTAL.

- see if the scancode map needs updating. But it's in one place! It was already not 1:1

**Comment by mcdurdin** [comment]
> ```
>   0x56: k.K_oE2, // << Dup
>   0x73: k.k_oC1,
>   0x7D: k.K_oE2, // << Dup
> ```
> 
> except we already have dups.

This would only be a problem if both `0x56` and `0x7D` scan codes appear on the same hardware keyboard, which they do not: `0x56` is on ISO and ABNT2, and `0x7D` is on JIS.

File: common/web/types/src/consts/virtual-key-constants.ts [hunk 9]
```diff
@@ -144,72 +144,92 @@ export const USVirtualKeyCodes = {
...
-  // [caps] A S D F G H J K L ç ~ ] [enter]
-  [ k.K_A, k.K_S, k.K_D, k.K_F, k.K_G, k.K_H, k.K_J, k.K_K, k.K_L, k.K_COLON, k.K_QUOTE, k.K_BKSLASH ],
-  // [shift] \ Z X C V B N M , . ; / [shift]
-  [ k.K_oE2, k.K_Z, k.K_X, k.K_C, k.K_V, k.K_B, k.K_N, k.K_M, k.K_COMMA, k.K_PERIOD, k.K_SLASH, k.k_oC1 /* ABNT2 */ ],
-  // space
-  [ k.K_SPACE ],
-];
+  0x56: k.K_oE2, // << Dup
+  0x73: k.k_oC1,
+  0x7D: k.K_oE2, // << Dup
```
Comment (mcdurdin) [inline common/web/types/src/consts/virtual-key-constants.ts:204]:
```suggestion
  0x56: k.K_oE2, // << Same as 0x7D; found on iso, abnt2
  0x73: k.k_oC1,
  0x7D: k.K_oE2, // << Same as 0x56; found on jis
```

We need to fix the casing of `k.k_oC1` to `k.K_oC1` but that can be a separate PR I guess.

File: developer/src/kmc-ldml/test/fixtures/basic.txt
Comment (mcdurdin) [inline developer/src/kmc-ldml/test/fixtures/basic.txt:1]: This is a little painful to edit but it's still far better than a hex blob -- I think it is worth the pain!

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by srl295** [comment]
> > ```
> >   0x56: k.K_oE2, // << Dup
> >   0x73: k.k_oC1,
> >   0x7D: k.K_oE2, // << Dup
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > except we already have dups.
> 
> 
> This would only be a problem if both `0x56` and `0x7D` scan codes appear on the same hardware keyboard, which they do not: `0x56` is on ISO and ABNT2, and `0x7D` is on JIS.

Maybe I should exclude this specific dup.

File: common/web/types/src/consts/virtual-key-constants.ts (hunk 9 above)
Comment (srl295) [reply common/web/types/src/consts/virtual-key-constants.ts]: someother test failure so i'll pull that in…

**Comment by mcdurdin** [comment]
> Maybe I should exclude this specific dup.

Sounds sensible -- then we can track if future dups appear.

File: common/web/types/test/ldml-keyboard/test-ldml-keyboard-xml-reader.ts [hunk 10]
```diff
@@ -129,3 +142,32 @@ describe('ldml keyboard xml reader tests', function () {
...
+    assert.sameDeepMembers(ckeys, [[49, undefined]]);
+    assert.equal(badScans.size, 1);
+    assert.sameDeepMembers(Array.from(badScans.values()), [0xFF]);
+  });
+  it('CLDRScanToUSVirtualKeyCodes should be 1:1', () => {
+    const vkeyToScan = new Map<number, number>();
+    // check all scancodes
+    for (let scan = 0; scan <= 0xFF; scan++) {
+      const vkey = CLDRScanToVkey(scan);
+      if (vkey !== undefined && vkey !== 226) {
```
Comment (mcdurdin) [inline common/web/types/test/ldml-keyboard/test-ldml-keyboard-xml-reader.ts:166]: Can we use the defined constant here and add a comment to explain why we ignore K_oE2? And in fact, can we be more explicit about the test -- scan should be either 0x56 or 0x7D:

```suggestion
      if ((scan === 0x56 || scan === 0x7D) && vkey === USVirtualKeyCodes.K_oE2) {
        // We allow dup of K_oE2 / 0x56 iso,abnt2 / 0x7D - jis
        continue;
      }
      if (vkey !== undefined) {
```

Comment (srl295) [reply common/web/types/test/ldml-keyboard/test-ldml-keyboard-xml-reader.ts]: sorry i missed your suggestion.  in any event, my code catches if some other scan code duplicates K_oE2

**Review by mcdurdin — APPROVED** [review]
LGTM.

File: common/web/types/test/ldml-keyboard/test-ldml-keyboard-xml-reader.ts (hunk 10 above)
Comment (mcdurdin) [reply common/web/types/test/ldml-keyboard/test-ldml-keyboard-xml-reader.ts]: Your changes look good.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.185-alpha](https://keyman.com/downloads/releases/alpha/17.0.185)
